### PR TITLE
feature/ARSN 298/addHeapDataStructure

### PR DESCRIPTION
**Pull request by alexanderchan-scality**
- ARSN-298: support `NewerNoncurrentVersions`
- ARSN-298: add Min/Max heap data structure

### CODE REVIEW

File: lib/algos/heap/Heap.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+export enum CompareResult {
+    LT = -1,
+    EQ = 0,
+    GT = 1,
+}
+
+export type CompareFunction = (x: any, y: any) => CompareResult;
+
+export class Heap {
```
Comment (ghost) [inline lib/algos/heap/Heap.ts:14]: I am curious, what was the motivation to implement a heap in Arsenal? Don't we want to reuse existing heaps from proven open source projects?
Also, I am not sure (yet) to understand how it links to the addition of `NewerNoncurrentVersions`, is it to keep track of the objects?

Comment (alexanderchan-scality) [reply lib/algos/heap/Heap.ts:14]: @williamlardier
not sure about introducing a new dependency for this;  the heap structure should be a relatively simple structure to not need to  do so.
yes. the heap will be used to keep track of the `latest` objects. In the case of `NewNoncurrentVersion`, a heap is used over a list/array due to the asynchronous methods needed to determine if a version needs to be expired which may not be executed in order.

File: lib/algos/heap/Heap.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    size: number;
+    _maxSize: number;
+    _order: HeapOrder;
+    _heap: any[];
+    _cmpFn: CompareFunction;
```
Comment (ghost) [inline lib/algos/heap/Heap.ts:19]: Do we want to enforce private members?

```suggestion
    #maxSize: number;
    #order: HeapOrder;
    #heap: any[];
    #cmpFn: CompareFunction;
```

**Review by ghost — APPROVED** [review]

File: lib/algos/heap/Heap.ts (hunk 1 above)
Comment (ghost) [reply lib/algos/heap/Heap.ts:14]: Thanks for the explanation!

**Review by nicolas2bert — APPROVED** [review]

**Pull request merged**